include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -1,204 +1,508 @@
+            if (__comp(__val2, __val1))
+            {
+                __rng3[__start3 + __i] = __val2;
+                if (++__start2 == __n2)
+                {
+                    //copying a residual of the first seq
+                    for (++__i; __i < __chunk && __start1 < __n1; ++__i, ++__start1)
+                        __rng3[__start3 + __i] = __rng1[__start1];
+                }
+            }
+            else
+            {
+                __rng3[__start3 + __i] = __val1;
+                if (++__start1 == __n1)
+                {
+                    //copying a residual of the second seq
+                    for (++__i; __i < __chunk && __start2 < __n2; ++__i, ++__start2)
+                        __rng3[__start3 + __i] = __rng2[__start2];
+                }
+            }
+        }
+    }
+}
+
+// Please see the comment for __parallel_for_submitter for optional kernel name explanation
+template <typename _IdType, typename _MergeKernelName>
+struct __parallel_merge_submitter;
+
+template <typename _IdType, typename... _MergeKernelName>
+struct __parallel_merge_submitter<_IdType, __internal::__optional_kernel_name<_MergeKernelName...>>
+{
+    template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Compare>
+    auto
+    operator()(_ExecutionPolicy&& __exec, _Range1&& __rng1, _Range2&& __rng2, _Range3&& __rng3, _Compare __comp) const
+    {
+        const _IdType __n1 = __rng1.size();
+        const _IdType __n2 = __rng2.size();
+        const _IdType __n = __n1 + __n2;
+
+        assert(__n1 > 0 || __n2 > 0);
+
+        _PRINT_INFO_IN_DEBUG_MODE(__exec);
+
+        // Empirical number of values to process per work-item
+        const std::uint8_t __chunk = __exec.queue().get_device().is_cpu() ? 128 : 4;
+
+        const _IdType __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
+
+        auto __event = __exec.queue().submit(
+            [&__rng1, &__rng2, &__rng3, __comp, __chunk, __steps, __n1, __n2](sycl::handler& __cgh) {
+                oneapi::dpl::__ranges::__require_access(__cgh, __rng1, __rng2, __rng3);
+                __cgh.parallel_for<_MergeKernelName...>(
+                    sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id) {
+                        const _IdType __i_elem = __item_id.get_linear_id() * __chunk;
+                        const auto __start = __find_start_point(__rng1, __rng2, __i_elem, __n1, __n2, __comp);
+                        __serial_merge(__rng1, __rng2, __rng3, __start.first, __start.second, __i_elem, __chunk, __n1,
+                                       __n2, __comp);
+                    });
+            });
+        // We should return the same thing in the second param of __future for compatibility
+        // with the returning value in __parallel_merge_submitter_large::operator()
+        return __future(__event, __result_and_scratch_storage_base_ptr{});

```

Comment:
It looks like that `class __future` supports only `sycl::buffer`  or `__result_and_scratch_storage` or ...  there is another general version of `__wait_and_get_value` method:

    template <typename _T>
```
    constexpr auto
    __wait_and_get_value(const _T& __val)
    {
        wait();
        return __val;
    }
```
What `future` instance does return in case of 
`return __future(__event, __result_and_scratch_storage_base_ptr{});` ?
`std::shared_ptr` on the dummy `__result_and_scratch_storage_base` instance?

Reply:
The answer is in real use case:
```C++
        __par_backend_hetero::__parallel_merge(_BackendTag{}, ::std::forward<_ExecutionPolicy>(__exec),
                                               __buf1.all_view(), __buf2.all_view(), __buf3.all_view(), __comp)
            .__deferrable_wait();
```
So we simple waits.
About your question - probably we will return shared-pointer (`__result_and_scratch_storage_base_ptr`) in my implementation.

Reply:
> It looks like that `class __future` supports only `sycl::buffer` or `__result_and_scratch_storage` or ... there is another general version of `__wait_and_get_value` method:
> 
> ```c++
>     template <typename _T>
>     constexpr auto
>     __wait_and_get_value(const _T& __val)
>     {
>         wait();
>         return __val;
>     }
> ```
> 
> What `future` instance does return in case of `return __future(__event, __result_and_scratch_storage_base_ptr{});` ? `std::shared_ptr` on the dummy `__result_and_scratch_storage_base` instance?

It should and really return `__result_and_scratch_storage_base_ptr{}` instance.

Reply:
@MikeDvorskiy , can I mask this discussion as resolved?